codegen/generate.go

```diff
@@ -0,0 +1,126 @@
+package codegen
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"path/filepath"
+
+	"github.com/hashicorp/hcl2/hclwrite"
+	ctyjson "github.com/zclconf/go-cty/cty/json"
+
+	"github.com/gruntwork-io/terragrunt/errors"
+	"github.com/gruntwork-io/terragrunt/util"
+)
+
+// An enum to represent valid values for if_exists
+type GenerateConfigExists int
+
+const (
+	ExistsError GenerateConfigExists = iota
+	ExistsSkip
+	ExistsOverwrite
+	ExistsUnknown
+)
+
+const (
+	ExistsErrorStr     = "error"
+	ExistsSkipStr      = "skip"
+	ExistsOverwriteStr = "overwrite"
+)
+
+// Configuration for generating code
+type GenerateConfig struct {
+	Path     string
+	IfExists GenerateConfigExists
+	Contents string
+}
+
+// WriteToFile will generate a new file at the given target path with the given contents. If a file already exists at
+// the target path, the behavior depends on the value of IfExists:
+// - if ExistsError, return an error.
+// - if ExistsSkip, do nothing and return
+// - if ExistsOverwrite, overwrite the existing file
+func WriteToFile(logger *log.Logger, basePath string, config GenerateConfig) error {
+	// Figure out thee target path to generate the code in. If relative, merge with basePath.
+	var targetPath string
+	if filepath.IsAbs(config.Path) {
+		targetPath = config.Path
+	} else {
+		targetPath = filepath.Join(basePath, config.Path)
+	}
+
+	targetFileExists := util.FileExists(targetPath)
+	if targetFileExists && config.IfExists == ExistsError {
+		return errors.WithStackTrace(GenerateFileExistsError{path: targetPath})
+	} else if targetFileExists && config.IfExists == ExistsSkip {
+		// Do nothing since file exists and skip was configured
+		logger.Printf("The file path %s already exists and if_exists for code generation set to \"skip\". Will not regenerate file.", targetPath)
+		return nil
+	} else if targetFileExists {
```

Comment:
NIT: technically, the `GenerateConfigExists` enum has another state: `ExistsUnknown`. In theory, the code should never get here with that value, but perhaps return an error if it does?

Reply:
Good idea. Done: f9d88fa